### DISCUSSION TITLE
PedalinoMini concept with BPI leaf ESP32 S3

### DISCUSSION THREAD

**Discussion by Syntax-Dexter** [alf45tar/PedalinoMini#454] [Show and tell]
Hello there,

This awesome project came by and now I'm planning to build my own PedalinoMini. 
Nice job @alf45tar! Respect for all the coding.

I want to use this controller to control my THR10II amp via the THR remote app on my phone. This must be via BLE midi.
Second to control Amplitube 5 standalone and/or a DAW with VST plugins. Cubase AI was included by the THR amp.

I have finished my first controller with an arduino uno and a midi shield. Time for the next step.

So I made a design for a real controller powered by PedalinoMini.
The design is based on the dimensions of the THR amp.
Furthermore I add the typical THR grid to the front of the controller with a led behind. To get some tube power effect.
The top view is a basis 10 button concept, with momentary soft switches accompanied by vector based guitar pics and a RGB led.
The backside has different connections, some are optional for me.
Antenna, DC power, Midi thru are not that important to me. But I want it to be possible direct or in the future.
USB out is for connecting an powering a device like my phone or a tablet. This is handy for controlling via the THR app. I also can use my phone for duplicating the computer screen via Spacedesk. So I have visual on Amplitube and/or the DAW. I tested this with my android phone and a old android 9" tablet. Works like a charm.
The box will probably be build with wood. Specials images and text could be lasered.

I ordered a BPI leaf ESP32 S3 and a oled display and a bunch of stuff. Waiting for delivery and a holiday ahead.

But what wil you think guys;
Midi DIN vs 1/8 or 1/4 jack for midi in/out?
5mm (adressable) leds in a led holder or 2812 LED with a lens/reflector?
Hardware button to change midi polarity? I'm a hardware man thus like buttons en lights. Probably I end up with one footpedal.
Are some boot and reset buttons necessary?
Any tips or most common fails tho avoid?
Some rotary encoders/ potentiometers? I t would be nice if I can skip through an dial in some settings in the THR remote app.
Is it possible to make the controller be recognized as an iRig Stomp for example? I've read that there is an iRig blueboard config. How does this works?

![PedalinoMidi Concept 3D](https://github.com/alf45tar/PedalinoMini/assets/137293236/764f04f2-98be-4293-a09d-ff1c54030dd8)
![Voetpedaal midi](https://github.com/alf45tar/PedalinoMini/assets/137293236/39f90c90-7f1c-4012-953a-32d88a584f69)

![PedalinoMini Concept 1](https://github.com/alf45tar/PedalinoMini/assets/137293236/fef215a8-0112-4c96-86a2-11030e5a7c7b)

![PedalinoMidi LED options](https://github.com/alf45tar/PedalinoMini/assets/137293236/5908aa81-0364-4532-a68f-59193d672270)

**Comment by jaylfc** [comment]
Looks awesome!

**Reply by Syntax-Dexter** [reply]
Thanks!

**Comment by Syntax-Dexter** [comment]
So, I've got the most parts deliverd.

I can say that the OLED with the SPD0301 ic works with SSD1306/SH1106 and U8g2 libraries.
I choose this OLED because 1.54" looks like the perfect size for my project and it has an certain thickness. I hope it will fit nice and flat  in the housing. Also the 5mm LEDs with the WS2812 ic build inside looks promising.

This is my first electronics project ever so hopefully I can fix this. I just started with prototyping.

To be continued

![20230912_205324](https://github.com/alf45tar/PedalinoMini/assets/137293236/512a2380-2419-45a8-af1c-f4f3af748221)
![20230914_172648](https://github.com/alf45tar/PedalinoMini/assets/137293236/ff4c9ad4-caa9-47a5-b0c7-53d46580a280)
![20230914_172703](https://github.com/alf45tar/PedalinoMini/assets/137293236/ad486dbe-2840-470c-ad7f-4b2dd9b86e83)
![20230914_172715](https://github.com/alf45tar/PedalinoMini/assets/137293236/533985e5-b586-4159-a5ad-bbfee2e0a626)

**Reply by Designbase** [reply]
Where did you get the display from? The housing seems perfect to attach to a enclosure

**Reply by Syntax-Dexter** [reply]
https://a.aliexpress.com/_msOESBc

**Comment by Wapata** [comment]
Hi !
So far it look really promising.
So here is my advice : make your first unit with strictly what you need (if no hardware midi is mandatory, skip them). It's cheap enough to make an other one if you want to do other things later.

I've made an all in one, and I see that it will not be easy to change Banks on the fly because they do very different things....

**Comment by Syntax-Dexter** [comment]
Up and running for testing.
Managed to build 3.1.9 firmware with VSC + Platformio and upload it via USB connection with the PedalinoMini installer.

![Schermafbeelding 2023-09-18 205403](https://github.com/alf45tar/PedalinoMini/assets/137293236/ac1617c2-8601-4bbf-9671-56d74508043e)

**Comment by Syntax-Dexter** [comment]
Testing and prototyping with the ESP32 board. ( 3.1.10 )
Last day's I added a 18650 battery + RGB 5V leds + boost convertor 5V + power on led.
On battery power there's no 5V on the 5V pin, only when USB powered. So to let the leds work with the 3.3V logic a boost convertor is needed. 

To figur out:
1. How to connect and config rotary encoder and potentiometer -> any tips?
2. Get the right battery voltage levels. These levels are not corresponding with my multimeter. 
Also when BATTERY_PIN is connected to battery + the battery indicator keeps blinking like charging.
4. Amount of connected LEDs are not changeable in WebUI. Only in pedalino.h ?
5. Custom image's possible on oled?
6. Options for 3D printing of the housing. I have an Ultimaker 2+ at work, but it has a limited build volume of 223 x 220 x 205 mm (8.7 x 8.6 x 8 inches) 
7. Better understanding different button connections, pinouts and configurations.

Pedalino.h config:
`#elif defined ARDUINO_BPI_LEAF_S3`
`#undef  LEDS`
`#define LEDS  4`
`#undef  PEDALS`
`#define PEDALS 9`
`const byte pinD[] = {GPIO_NUM_1,  GPIO_NUM_2,  GPIO_NUM_3,  GPIO_NUM_4,  GPIO_NUM_5,  GPIO_NUM_6,  GPIO_NUM_7,  GPIO_NUM_8,  GPIO_NUM_0};`
`const byte pinA[] = {GPIO_NUM_11, GPIO_NUM_12, GPIO_NUM_13, GPIO_NUM_14, GPIO_NUM_46, GPIO_NUM_47, GPIO_NUM_17, GPIO_NUM_18, GPIO_NUM_0};`
`#define FACTORY_DEFAULT_PIN   GPIO_NUM_0    // Button BOOT`
`#define DIN_MIDI_IN_PIN       GPIO_NUM_44`
`#define DIN_MIDI_OUT_PIN      GPIO_NUM_43`
`#define BATTERY_PIN           GPIO_NUM_14   // Pin connected to +BATT`
`#define FASTLEDS_DATA_PIN     GPIO_NUM_48`
`#undef  SDA`
`#undef  SCL`
`#define SDA           GPIO_NUM_15 // standard I2C pin of onboard connector`
`#define SCL           GPIO_NUM_16 // standard I2C pin of onboard connector`


https://github.com/alf45tar/PedalinoMini/assets/137293236/ddd68a1f-7250-4e48-95de-d9d449f19d67

![20230927_180053](https://github.com/alf45tar/PedalinoMini/assets/137293236/5c25fd15-ba50-421e-a2c9-56e7bd299e1c)

**Reply by regisampa** [reply]
> To figur out:
> 

> 2. Get the right battery voltage levels. These levels are not corresponding with my multimeter.
>    Also when BATTERY_PIN is connected to battery + the battery indicator keeps blinking like charging.

Hello friend, you managed to solve the problem of indicating the battery charge? I am in the same situation, the battery indicator keeps flashing as if it is charging and does not show the real charge value, I use a 18650 battery. Everything works perfectly on the pedalboard but the battery indication does not work. Thank you very much

**Reply by Syntax-Dexter** [reply]
I am going to try to use a different pin for battery level read. As pin 14 is the _default_ battery pin, but also an ADC2 pin that has documented conflicts with active Wi-Fi.

**Comment by Wapata** [comment]
Point number five : https://formlabs.com/blog/how-to-create-models-larger-than-your-3d-printers-build-volume/
;-)

**Comment by jisv48** [comment]
How will you be charging the battery? are you incorporating a charger for the Lithium batt?  Your project looks great, by the way what happen with the Pedalino case? who will be selling them?  Cheers

**Reply by Syntax-Dexter** [reply]
There is a charging ic /circuit onboard

**Comment by Syntax-Dexter** [comment]
@alf45tar 

I am still learning all the possibilities in the WebUI to control Amplitube and Yamaha THR remote App. Both programs use CC commands. Although Amplitube has PC commands only for presets selection (1 - 128)

I can set on/off to pedals or changing knobs with an potentiometer. So far so good.

What I didn't can is implement and use a rotary encoder (20 pulses). How to connect these 3 pins? (Pin A,B and button pin) 
The following site explains how to use rotary encoders and only use 1 analog pin. Perhaps you already have some thoughts or experience with this?
[https://www.elektormagazine.nl/labs/rotary-encoder-on-a-single-mcu-pin](url)

Is it some way possible to use 1 or 2 rotary encoders or potentiometers and knobs to assign midi cc commands to THR remote APP?
So different CC commands for applying on/off effects , adjust values, toggle through cabinets or effects -> toggle  spring/room/plate/hall for reverb for example. It would be nice if the PedalinoMini can control al the parameters with only 1 or 2 knobs/encoders. So you can dial in some tones on the PedalinoMini instead of using a PC . Using a mouse for rotating knobs is not so fine.
There are so many functions, pedals, control, actions and sequences. But get them programmed well is another story.
Perhaps you can give me an direction how to use all the options in the right manner.

Thank you in advance



![Screenshot_20231013_194656_THR Remote](https://github.com/alf45tar/PedalinoMini/assets/137293236/bcb5d9cc-2e9b-446a-8d27-a920580f04e5)

**Comment by alf45tar** [comment]
Rotary encoders are supported since long time by PedalinoMini. They are called Jog wheel. Check the schematic for connection.

**Comment by jisv48** [comment]
Hi alf45Star,

What is the status of Enclosures for the original Pedalino, are you going to sell them? if so how do we order them and the cost per unit?

Thanks in advance for your assistance

Cheers

**Reply by alf45tar** [reply]
No plan. Do you need the 3D printed case or all the components too?

**Comment by jisv48** [comment]
The 3D printed case. I have already assembled the PedalinoMini on a breadboard, it will be easy to move the parts there.  Thanks

**Reply by alf45tar** [reply]
Did you tried to get a quotation from an online 3D printing service?

**Comment by jisv48** [comment]
Well, I probably did, but that was long ago.  I’ll look into that.  Thanks

**Reply by alf45tar** [reply]
I can print for you for 30 EUR + shipping but I prefers you go to a professional 3D printing service.

**Comment by Syntax-Dexter** [comment]
I designed a smaller version 200x115mm. today. 

![Mini and Maxi rev1](https://github.com/alf45tar/PedalinoMini/assets/137293236/8fa2ab57-aced-40ea-9e27-6f9bfe9570f6)

**Comment by Syntax-Dexter** [comment]
@alf45tar 
Do you have tips for setting up 3D printing  / CURA in general? 
I assume you print your PedalinoMini body upside down. Do you use any sort of support type or do you use infills?
What is you're experience with wall thickness, strength and warping? Case1 looks very robust!

**Comment by alf45tar** [comment]
I printed it upside down of course. Supports are requested for the ceiling of ESP32 slot.
I am using low quality setting of CURA and they are enough for me. The case is very robust with 20% of infill.

**Reply by Syntax-Dexter** [reply]
Thanks!

**Comment by Syntax-Dexter** [comment]
Raining day........
Update design

![PedalinMini Rev2 2](https://github.com/alf45tar/PedalinoMini/assets/137293236/244157f9-d397-4068-905e-3a51dec13c67)
![PedalinMini Rev2 3](https://github.com/alf45tar/PedalinoMini/assets/137293236/54ee15b7-ce18-4535-a8d2-5f94bfe71470)
![PedalinMini Rev2 1](https://github.com/alf45tar/PedalinoMini/assets/137293236/1c865b09-7819-4c27-adff-508053de1fff)
![PedalinMini Rev2 4](https://github.com/alf45tar/PedalinoMini/assets/137293236/1fdd381c-4483-4e43-92d2-4acef8e2fc52)

**Comment by Syntax-Dexter** [comment]
Black Friday deals, I ordered a Bambu Lab P1S printer for the Christmas holiday. So now time to learn more about support settings and so on. And make a choice for filament and color.
And making up an excuse to my wife, ADHD will not be enough 😆

**Comment by Syntax-Dexter** [comment]
I printed the enclosure last week and fitted the first parts. Everything looks good and the dimensions are perfect. Hopefully I can soon soldering it all together.  
You see at the back tactile buttons for reset and boot, switch to disconnect the battery, USB C, 2x TRS 1/4" and 2x female DIN 5. Screen is 1.54 OLED and 8x momentary footswitches. 
To be continued....

![20240111_220131](https://github.com/alf45tar/PedalinoMini/assets/137293236/27cde9c3-16ed-4ce2-b53d-1cc7fd52cd59)
![20240111_220518](https://github.com/alf45tar/PedalinoMini/assets/137293236/a406c9c0-85ea-4de2-8c7b-98e684c2ecfc)
![20240112_091453](https://github.com/alf45tar/PedalinoMini/assets/137293236/e06b5c66-52be-47ee-82e7-aedf35f416ac)

**Comment by jisv48** [comment]
Excellent job, looks great. congratulations.

**Comment by Syntax-Dexter** [comment]
@alf45tar 
I build with the latest firmware 3.2.1 via the webtools installer. (BPI Leaf S3)
Config for now is: (pedal 5 and 6 set to none in the WebUI)

![Schermafbeelding 2024-02-01 143401](https://github.com/alf45tar/PedalinoMini/assets/137293236/0b95fe02-ba20-43ea-a75b-dce70b2ee290)

I can't use A14 as battery_pin and analog_pin for pedal 4 at the same time. The button gives no response, as it has 2 analog functions?
Also pin 15/16 the SDA and SCL pins are also set to A5 and A6. Simple solution to free ADC pins from pinD array? And put non ADC GPIO's  there instead?  > Update: looks to work with: D4(21) A4(7), D5(35) A5(6), D6(36) A6(5), <
Log from Pedal 4;

![Schermafbeelding 2024-02-01 143414](https://github.com/alf45tar/PedalinoMini/assets/137293236/192792cd-1f21-4d96-a8b7-ba413df34a75)


So I wonder if pins 35-42 are for example usable?
I have BPI Leaf S3 and it has Quad Flash and not Octa Flash, wich could add the use of pins 33-37
For as far I understand USB serial/JTAG only needs 19, 20 for D- and D+ and GND + Vcc.
So in my theory without many knowledge of EPS32 pins, all did I read some documentation.
Complete rearrangement of pin settings with a total of 13 pedals, Analog pins are ADC1 or ADC2. 
Midi TX/RX to pin 21 and 47.
I did not used pin 3, 14, 15, 16, 19, 20, 45, 46. 
Are these pin settings possible? That would give a lot more options.

`const byte pinD[] = {GPIO_NUM_41, GPIO_NUM_42,  GPIO_NUM_43,  GPIO_NUM_44,  GPIO_NUM_35,  GPIO_NUM_36,  GPIO_NUM_37,  GPIO_NUM_38,  GPIO_NUM_39, GPIO_NUM_40, GPIO_NUM_17, GPIO_NUM_18, GPIO_NUM_0};`

`const byte pinA[] = {GPIO_NUM_11, GPIO_NUM_12,  GPIO_NUM_13,  GPIO_NUM_4,   GPIO_NUM_5,   GPIO_NUM_6,   GPIO_NUM_7,   GPIO_NUM_8,   GPIO_NUM_9,  GPIO_NUM_10, GPIO_NUM_1,  GPIO_NUM_2,  GPIO_NUM_0};`


The goal for me now is 8 Momentary_1 pedals + 2x Analog TRS inputs.
And it would be nice if SLOTS are freely assignable, or at least to different buttons in the same pedal.
With 4 x Momentary_2 only the fist button can be in the SLOT name. #434 
And a possibility to chose between 3 or 4 SLOT_COLS would be nice also.

**Reply by alf45tar** [reply]
What do you think to define like of this?
````
#elif defined ARDUINO_BPI_LEAF_S3           // https://wiki.banana-pi.org/BPI-Leaf-S3
#undef  PEDALS
#define PEDALS                9
// https://docs.espressif.com/projects/esp-idf/en/latest/esp32s3/api-reference/peripherals/gpio.html
// Strapping pins: 1, 3, 45, 46
// ADC1:    1-10
// ADC2:   11-20 (19=D- 20=D+) (ADC2 module is also used by the Wi-Fi)
// SPI0/1: 26-32 (usually used for SPI flash and PSRAM)
// SPI0/1: 33-37 (on boards embedded with ESP32-S3R8 / ESP32-S3R8V chip, GPIO33~37 are also not recommended for other uses)
const byte pinD[] = {GPIO_NUM_38, GPIO_NUM_39, GPIO_NUM_40, GPIO_NUM_41, GPIO_NUM_42, GPIO_NUM_43, GPIO_NUM_44, GPIO_NUM_2, GPIO_NUM_0};
const byte pinA[] = {GPIO_NUM_4,  GPIO_NUM_5,  GPIO_NUM_6,  GPIO_NUM_7,  GPIO_NUM_8,  GPIO_NUM_9,  GPIO_NUM_10, GPIO_NUM_2, GPIO_NUM_0};
#define FACTORY_DEFAULT_PIN   GPIO_NUM_0    // Button BOOT
#define DIN_MIDI_IN_PIN       GPIO_NUM_21
#define DIN_MIDI_OUT_PIN      GPIO_NUM_47
#define BATTERY_PIN           GPIO_NUM_14   // Pin connected to +BATT
#define FASTLEDS_DATA_PIN     GPIO_NUM_48
#undef  SDA
#undef  SCL
#define SDA                   GPIO_NUM_15
#define SCL                   GPIO_NUM_16
````

**Reply by Syntax-Dexter** [reply]
Hi Alf45tar,
That code would work, although A8 GPIO_NUM_2  can be set to GPIO_NUM_11

I did tried the code below already for a **9 pedal** version. That works fine as far I can tell. I think using ADC2 pin 35-37 as digital input is not a problem. All the buttons works as to be aspected in a 4x Momentary_2 + 1x Jogwheel breadboard testbuild.

I also introduced a **13 pedal** alternative, the Pedalino**XTREME**  :sunglasses:  :metal:
Not tested yet, but it has potential. For example, 13 controller ports x 6 switches ladder = 78 switches or 13 rotary encoders.
You will need an extra leg or band member to master the PedalinoXTREME :yum:

Even strapping pins 45 and 46 can be added as digital I/O pin as long as they are LOW at startup/after reset.
_All strapping pins have latches. At system reset, the latches sample the bit values of their respective strapping
pins and store them until the chip is powered down or shut down. The states of latches cannot be changed in
any other way. It makes the strapping pin values available during the entire chip operation, **and the pins are freed
up to be used as regular IO pins after reset.**
Regarding the timing requirements for the strapping pins, there are such parameters as setup time and hold time._

```c
// --------------------------------------------------------------------------
#elif defined ARDUINO_BPI_LEAF_S3           // https://wiki.banana-pi.org/BPI-Leaf-S3
#undef  PEDALS
#define PEDALS                13
// https://docs.espressif.com/projects/esp-idf/en/latest/esp32s3/api-reference/peripherals/gpio.html
// Strapping pins : 1, 3, 45, 46
// Default pins   : 14 to +BATT, 15 SDA, 16 SCL, 0 BOOT, 3 RESET
// USB JTAG : 19=D-, 20=D+
// ADC1     :  1-10
// ADC2     : 11-20 (Espressif 4.x ADC2 has hardware limitations when also used by Wifi, usable between wifi(stop) and wifi(start)
// ADC2     : 11-20 (Espressif 5.x introduced oneshot_read instead of continious_read as workaround)
// SPI0/1   : 26-32 (NOT recommended, used for SPI flash and PSRAM)
// SPI0/1   : 33-37 (BPI LEAF S3 is not embedded with ESP32-S3R8 / ESP32-S3R8V chip, GPIO33~37 are fine for input pins) *note; pin 33 and 34 are not present

// ** PedalinoMini STANDARD: 9 PEDALS **
// Shown in WebUI as | Pedal 1,   | Pedal 2,   | Pedal 3,   | Pedal 4,   | Pedal 5,   | Pedal 6,   | Pedal 7,  |  Pedal 8,  | Pedal 9, | 
const byte pinD[] = {GPIO_NUM_10, GPIO_NUM_9,  GPIO_NUM_8,  GPIO_NUM_21, GPIO_NUM_35, GPIO_NUM_36, GPIO_NUM_4,  GPIO_NUM_2,  GPIO_NUM_0, GPIO_NUM_0, GPIO_NUM_0, GPIO_NUM_0, GPIO_NUM_0};
const byte pinA[] = {GPIO_NUM_11, GPIO_NUM_12, GPIO_NUM_13, GPIO_NUM_7,  GPIO_NUM_6,  GPIO_NUM_5,  GPIO_NUM_17, GPIO_NUM_18, GPIO_NUM_0, GPIO_NUM_0, GPIO_NUM_0, GPIO_NUM_0, GPIO_NUM_0};

// ** ALTERNATIVE the PedalinoXTREME: 13 PEDALS **
//Shown in WebUI as   | Pedal 1,   | Pedal 2,   | Pedal 3,  |  Pedal 4,   |  Pedal 5,   |  Pedal 6,   |  Pedal 7,   | Pedal 8,   | Pedal 9,   | Pedal 10   | Pedal 11   | Pedal 12   | Pedal 13  | 
//const byte pinD[] = {GPIO_NUM_41, GPIO_NUM_42, GPIO_NUM_43, GPIO_NUM_44,  GPIO_NUM_35,  GPIO_NUM_36,  GPIO_NUM_37, GPIO_NUM_38, GPIO_NUM_39, GPIO_NUM_40, GPIO_NUM_17, GPIO_NUM_18, GPIO_NUM_0};
//const byte pinA[] = {GPIO_NUM_11, GPIO_NUM_12, GPIO_NUM_13, GPIO_NUM_4,   GPIO_NUM_5,   GPIO_NUM_6,   GPIO_NUM_7,  GPIO_NUM_8,  GPIO_NUM_9,  GPIO_NUM_10, GPIO_NUM_1,  GPIO_NUM_2,  GPIO_NUM_0};

#define FACTORY_DEFAULT_PIN   GPIO_NUM_0      // Button BOOT
#define DIN_MIDI_IN_PIN       GPIO_NUM_44     // (RX, if PedalinoMini STANDARD)
#define DIN_MIDI_OUT_PIN      GPIO_NUM_43     // (TX, if PedalinoMini STANDARD)
//#define DIN_MIDI_IN_PIN       GPIO_NUM_21     // (RX, if PedalinoMini XTREME)
//#define DIN_MIDI_OUT_PIN      GPIO_NUM_47     // (TX, if PedalinoMini XTREME)
#define BATTERY_PIN           GPIO_NUM_14     // Pin connected to +BATT, connected to onboard MX 1.25mm 2-Pin Socket
#define FASTLEDS_DATA_PIN     GPIO_NUM_48
#undef  SDA
#undef  SCL
#define SDA                   GPIO_NUM_15   // Default pin, also connected to onboard JST SH 1mm 4-Pin Socket (I2C), if you don't use the socket feel free to change
#define SCL                   GPIO_NUM_16   // Default pin, also connected to onboard JST SH 1mm 4-Pin Socket (I2C), if you don't use the socket feel free to change
// --------------------------------------------------------------------------
````
I think schema.json needs to be updated to accept more pedals?
https://github.com/alf45tar/PedalinoMini/blob/a801ccfee86c0a3037d9ee00b4d81f025c39006e/data/schema.json#L405
https://github.com/alf45tar/PedalinoMini/blob/a801ccfee86c0a3037d9ee00b4d81f025c39006e/data/schema.json#L411
https://github.com/alf45tar/PedalinoMini/blob/a801ccfee86c0a3037d9ee00b4d81f025c39006e/data/schema.json#L467
https://github.com/alf45tar/PedalinoMini/blob/a801ccfee86c0a3037d9ee00b4d81f025c39006e/data/schema.json#L475

**Comment by Syntax-Dexter** [comment]
By the way, ADC2 pins 11-20 can be used for digital inputs/outputs. Only analog use of these pins conflicts with Wifi enabled. Wifi is also on the analog internal circuitry.

**Comment by Syntax-Dexter** [comment]
I build another PedalinoMini :-)
This time in black matte PLA. Still 8 footswitches, 2x TRS and 1x DIN.
For better battery voltage measuring I used gpio 1 with a voltage divider. This works better for me then the default gpio 14 on the BPI-leaf-S3. (ADC2 pin's doesn't do analog conversion with WiFi enabled)
In addition I added a second ledstrip for atmospheric lightning to match my Yamaha THR amp.
Display is ST7789 240 x 240 pixels, SPI connection. 
I used older 3.5.0 firmware and tweaked display, extra atmospheric lightning and other battery voltage pin.
Works like a charm.

![20260326_140503](https://github.com/user-attachments/assets/790895dd-3141-409a-863b-e7c434d1830f)
![20260326_141723](https://github.com/user-attachments/assets/02801cbc-32b8-4671-8e91-3ce022861e4a)
![20260326_140057](https://github.com/user-attachments/assets/06c130ff-e4c1-4f10-8e70-dde4ae1073b2)
![20260326_140406](https://github.com/user-attachments/assets/7d7f1fe1-b31f-4cb7-871e-95aa4a4ee2d0)
![20260326_140529](https://github.com/user-attachments/assets/f7895149-a03e-4c4f-8091-79ca1c71f82b)

**Reply by alf45tar** [reply]
Awesome

**Reply by maarthome** [reply]
> I build another PedalinoMini :-) This time in black matte PLA. Still 8 footswitches, 2x TRS and 1x DIN. For better battery voltage measuring I used gpio 1 with a voltage divider. This works better for me then the default gpio 14 on the BPI-leaf-S3. (ADC2 pin's doesn't do analog conversion with WiFi enabled) In addition I added a second ledstrip for atmospheric lightning to match my Yamaha THR amp. Display is ST7789 240 x 240 pixels, SPI connection. I used older 3.5.0 firmware and tweaked display, extra atmospheric lightning and other battery voltage pin. Works like a charm.
> 
> ![20260326_140503](https://private-user-images.githubusercontent.com/137293236/569793556-790895dd-3141-409a-863b-e7c434d1830f.jpg?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzQ2MDA2NTIsIm5iZiI6MTc3NDYwMDM1MiwicGF0aCI6Ii8xMzcyOTMyMzYvNTY5NzkzNTU2LTc5MDg5NWRkLTMxNDEtNDA5YS04NjNiLWU3YzQzNGQxODMwZi5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMzI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDMyN1QwODMyMzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0wYjJjNDYyNDBmODgzNGVlNzExYjE2NTNmMmUwZmQ0NGQ2ODVjYmY3M2VhMThiYTQzODc4Y2M5N2ViMzU3OWE5JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.jlqEnRZFS-JBnA21vbBvivmU0cH3HMX5O3cQePmFisw) ![20260326_141723](https://private-user-images.githubusercontent.com/137293236/569793575-02801cbc-32b8-4671-8e91-3ce022861e4a.jpg?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzQ2MDA2NTIsIm5iZiI6MTc3NDYwMDM1MiwicGF0aCI6Ii8xMzcyOTMyMzYvNTY5NzkzNTc1LTAyODAxY2JjLTMyYjgtNDY3MS04ZTkxLTNjZTAyMjg2MWU0YS5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMzI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDMyN1QwODMyMzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT00Njk0MjIzMDUwM2RjNWRkNWFhYTlmZmNlMzQwYTVlMDVhNDhmNWIzMzgxYzk3ZGI1ZGU0ZjA4MDQxYWZiNmE1JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.-OX40mUpELaM_Lhiq_qj5sDp3B8uglU6nbsZoAi4D04) ![20260326_140057](https://private-user-images.githubusercontent.com/137293236/569793592-06c130ff-e4c1-4f10-8e70-dde4ae1073b2.jpg?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzQ2MDA2NTIsIm5iZiI6MTc3NDYwMDM1MiwicGF0aCI6Ii8xMzcyOTMyMzYvNTY5NzkzNTkyLTA2YzEzMGZmLWU0YzEtNGYxMC04ZTcwLWRkZTRhZTEwNzNiMi5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMzI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDMyN1QwODMyMzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1hMWVmMTkwYTQ5MzdmNTc1NGU0MWVhOTdmZTY0N2FkYTI3NzM1ZGM0NzEzNjA1NTAzMGE4MjE2YWZiNGQyMWQ0JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.tVx2MuwdBsEP4UNgIGEDHVIr6m1TvWIG4KKOZhronFk) ![20260326_140406](https://private-user-images.githubusercontent.com/137293236/569793624-7d7f1fe1-b31f-4cb7-871e-95aa4a4ee2d0.jpg?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzQ2MDA2NTIsIm5iZiI6MTc3NDYwMDM1MiwicGF0aCI6Ii8xMzcyOTMyMzYvNTY5NzkzNjI0LTdkN2YxZmUxLWIzMWYtNGNiNy04NzFlLTk1YWE0YTRlZTJkMC5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMzI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDMyN1QwODMyMzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jNzRkMmE3NzQ5OTRmNTJkNTY5Nzg0ZDg1NzA4MDM3ODg1NzVjZjY1N2Y3ZjU0N2M5MzQzNzM0NWQwMDkzZmJkJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.acNxNrsAdaz7yKw4Emr2Vld46ytMDDFxQFPyTGi5Sdw) ![20260326_140529](https://private-user-images.githubusercontent.com/137293236/569793647-f7895149-a03e-4c4f-8091-79ca1c71f82b.jpg?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzQ2MDA2NTIsIm5iZiI6MTc3NDYwMDM1MiwicGF0aCI6Ii8xMzcyOTMyMzYvNTY5NzkzNjQ3LWY3ODk1MTQ5LWEwM2UtNGM0Zi04MDkxLTc5Y2ExYzcxZjgyYi5qcGc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjYwMzI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI2MDMyN1QwODMyMzJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1mMThjMjg2ZDg0MjIwYzg1YTUxNWVkZmFiZTZkNzUwNDNkNGU2YzdiNWMyYWNlMWMwNjBmYmFjNjg1YjNiMjA4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.2gofy2prKKYyD8C6hhPbMueOy34MHvR49Qsw0U9sogk)

Hi @Syntax-Dexter ,
Wow! Looks great your pedal! Super!


Is your display adaption code available ? If so, where?

Cheers and enoy your day,
Maarten

**Reply by Syntax-Dexter** [reply]
Do you have BPI + 4 SPI wire ST7789 320x240 2" display?
Anyway, I have a fork of PedalinoMini 3.5.0 in my repositories with the code changes for this display and battery pin.

**Reply by maarthome** [reply]
Thanks  a lot!
and yes I do have them.

Cheers,
Maarten

**Reply by rsieiro** [reply]
Hey @Syntax-Dexter! I'm planning to do something similar to what you did. Is the LCD display required or could I go display-less if I wanted to? In case a LCD display is required, would any ST7789 with SPI work? Like [this one on AliExpress](https://pt.aliexpress.com/item/1005012270101602.html), for example?
